From 1533c5ba1e2d0b76269dc58d4eb52a55c0ac0051 Mon Sep 17 00:00:00 2001
From: Yi Zhao <yi.zhao@windriver.com>
Date: Sun, 30 Jun 2024 22:24:57 +0800
Subject: [PATCH] policy/modules/system/userdomain: allow administrative user
 to get attributes of shadow history file

Before the patch:
root@qemux86-64:~# ls -lZ /etc/security/opasswd
-?????????? ? ?    ?    ?                                    ?  ? /etc/security/opasswd

After the patch:
root@qemux86-64:~# ls -lZ /etc/security/opasswd
-rw-------. 1 root root user_u:object_r:shadow_history_t 237 Jun 30 12:03 /etc/security/opasswd

Upstream-Status: Submitted [https://github.com/SELinuxProject/refpolicy/pull/789]

Signed-off-by: Yi Zhao <yi.zhao@windriver.com>
---
 policy/modules/system/authlogin.if  | 19 +++++++++++++++++++
 policy/modules/system/userdomain.if |  1 +
 2 files changed, 20 insertions(+)

diff --git a/policy/modules/system/authlogin.if b/policy/modules/system/authlogin.if
index 831c5275b..15f2e1267 100644
--- a/policy/modules/system/authlogin.if
+++ b/policy/modules/system/authlogin.if
@@ -742,6 +742,25 @@ interface(`auth_etc_filetrans_shadow',`
 	files_etc_filetrans($1, shadow_t, file, $2)
 ')
 
+########################################
+## <summary>
+##	Get the attributes of the shadow history file.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`auth_getattr_shadow_history',`
+	gen_require(`
+		type shadow_history_t;
+	')
+
+	files_search_etc($1)
+	allow $1 shadow_history_t:file getattr;
+')
+
 ########################################
 ## <summary>
 ##	Read the shadow history file.
diff --git a/policy/modules/system/userdomain.if b/policy/modules/system/userdomain.if
index a173cb282..837bb49e3 100644
--- a/policy/modules/system/userdomain.if
+++ b/policy/modules/system/userdomain.if
@@ -1426,6 +1426,7 @@ template(`userdom_admin_user_template',`
 	term_use_all_terms($1_t)
 
 	auth_getattr_shadow($1_t)
+	auth_getattr_shadow_history($1_t)
 	# Manage almost all files
 	files_manage_non_auth_files($1_t)
 	files_map_non_auth_files($1_t)
-- 
2.25.1

